### ISSUE TITLE
Failure to login using Active Directory (Windows & Samba)

### ISSUE THREAD

**Issue by alexanderharm** [nuxsmin/sysPass#1437]
**sysPass Version**
3.1 (310.19081801)

**Describe the question**
I'm using the Docker version of sysPass and I have a hard time getting LDAP auth working. I configured LDAP in the config interface and I am able to import users & groups. I tried with sAMAccountName and userPrincipalName. However, once I enable LDAP and go to the login page the imported users are unable to login in.
If I login into the Docker image ldapsearch works just fine with the user & password I try to use to login (again both via sAMAccountName and userPrincipalName).

So to summarize:
- LDAP test on config page: works
- LDAP user import: works
- LDAP group import: works
- LDAP CLI: works
- LDAP auth: fails

**Platform (please complete the following information):**
 - OS: Linux
 - Debian Stretch
 - Firefox, Brave

**Additional context**
Using Docker image.
Using self-signed certificates, the root cert has been added to `/etc/ssl/certs/ca-certificates.crt` via `update-ca-certificates`. `TLS_REQCERT allow` set in `/etc/openldap/ldap.conf`.
Using authenticator plugin.

**Comment by nuxsmin** [comment]
Hello, could you check out for any error messages within `.../app/config/syspass.log` file?

Regards

**Comment by alexanderharm** [comment]
```
[2019-09-02 23:13:07] syspass.EXCEPTION: logger {"message":"LDAP-Verbindungsfehler (BIND)
#0 /var/www/html/sysPass/lib/SP/Providers/Auth/Ldap/LdapConnection.php(114): SP\Providers\Auth\Ldap\LdapConnection->bind()
#1 /var/www/html/sysPass/lib/SP/Providers/Auth/Ldap/LdapConnection.php(95): SP\Providers\Auth\Ldap\LdapConnection->connectAndBind()
#2 /var/www/html/sysPass/lib/SP/Providers/Auth/Ldap/Ldap.php(96): SP\Providers\Auth\Ldap\LdapConnection->checkConnection()
#3 /var/www/html/sysPass/lib/SP/Providers/Auth/AuthProvider.php(154): SP\Providers\Auth\Ldap\Ldap->factory(Object(SP\Providers\Auth\Ldap\LdapParams),Object(SP\Core\Events\EventDispatcher),Boolean)
#4 /var/www/html/sysPass/lib/SP/Providers/Auth/AuthProvider.php(116): SP\Providers\Auth\AuthProvider->getLdapAuth()
#5 /var/www/html/sysPass/lib/SP/Providers/Auth/AuthProvider.php(97): SP\Providers\Auth\AuthProvider->authLdap()
#6 /var/www/html/sysPass/lib/SP/Services/Auth/LoginService.php(154): SP\Providers\Auth\AuthProvider->doAuth(Object(SP\DataModel\UserLoginData))
#7 /var/www/html/sysPass/app/modules/web/Controllers/LoginController.php(65): SP\Services\Auth\LoginService->doLogin()
#8 [internal function]: SP\Modules\Web\Controllers\LoginController->loginAction()
#9 /var/www/html/sysPass/lib/SP/Bootstrap.php(240): call_user_func_array(Array,Array)
#10 [internal function]: SP\Bootstrap->SP\{closure}(Object(Klein\Request),Object(Klein\Response),Object(Klein\ServiceProvider),Object(Klein\App),Object(Klein\Klein),Object(Klein\DataCollection\RouteCollection),Array)
#11 /var/www/html/sysPass/vendor/klein/klein/src/Klein/Klein.php(886): call_user_func(Object(Closure),Object(Klein\Request),Object(Klein\Response),Object(Klein\ServiceProvider),Object(Klein\App),Object(Klein\Klein),Object(Klein\DataCollection\RouteCollection),Array)
#12 /var/www/html/sysPass/vendor/klein/klein/src/Klein/Klein.php(588): Klein\Klein->handleRouteCallback(Object(Klein\Route),Object(Klein\DataCollection\RouteCollection),Array)
#13 /var/www/html/sysPass/lib/SP/Bootstrap.php(464): Klein\Klein->dispatch(Object(Klein\Request))
#14 /var/www/html/sysPass/lib/Base.php(75): SP\Bootstrap->run(Object(DI\Container))
#15 /var/www/html/sysPass/index.php(28): require(String)","caller":"N/A"}
```

I double and triple checked again. I have an OPNsense (also PHP) here that works perfectly well with the same credentials...

**Comment by nuxsmin** [comment]
Hello, could you enable LDAP logging messages by adding  `ldap.bind` as event on logging configuration?

Regards

**Comment by alexanderharm** [comment]
How?

Like this:

```
 <logEnabled>1</logEnabled>
 <logEvents>ldap.bind</logEvents>
```

? Which gives no different output...

**Comment by nuxsmin** [comment]
It can be do through the web UI. It should log some events in the event log (web).

**Comment by nuxsmin** [comment]
All the XMl configuration can be done through the web UI, it very advisable to do not modify it by hand unless you can't access to the web UI

**Comment by alexanderharm** [comment]
Makes not much difference:

```
[2019-09-03 18:36:56] syspass.EXCEPTION: logger {"message":"LDAP-Verbindungsfehler (BIND)
#0 /var/www/html/sysPass/lib/SP/Providers/Auth/Ldap/LdapConnection.php(114): SP\Providers\Auth\Ldap\LdapConnection->bind()
#1 /var/www/html/sysPass/lib/SP/Providers/Auth/Ldap/LdapConnection.php(95): SP\Providers\Auth\Ldap\LdapConnection->connectAndBind()
#2 /var/www/html/sysPass/lib/SP/Providers/Auth/Ldap/Ldap.php(96): SP\Providers\Auth\Ldap\LdapConnection->checkConnection()
#3 /var/www/html/sysPass/lib/SP/Providers/Auth/AuthProvider.php(154): SP\Providers\Auth\Ldap\Ldap->factory(Object(SP\Providers\Auth\Ldap\LdapParams),Object(SP\Core\Events\EventDispatcher),Boolean)
#4 /var/www/html/sysPass/lib/SP/Providers/Auth/AuthProvider.php(116): SP\Providers\Auth\AuthProvider->getLdapAuth()
#5 /var/www/html/sysPass/lib/SP/Providers/Auth/AuthProvider.php(97): SP\Providers\Auth\AuthProvider->authLdap()
#6 /var/www/html/sysPass/lib/SP/Services/Auth/LoginService.php(154): SP\Providers\Auth\AuthProvider->doAuth(Object(SP\DataModel\UserLoginData))
#7 /var/www/html/sysPass/app/modules/web/Controllers/LoginController.php(65): SP\Services\Auth\LoginService->doLogin()
#8 [internal function]: SP\Modules\Web\Controllers\LoginController->loginAction()
#9 /var/www/html/sysPass/lib/SP/Bootstrap.php(240): call_user_func_array(Array,Array)
#10 [internal function]: SP\Bootstrap->SP\{closure}(Object(Klein\Request),Object(Klein\Response),Object(Klein\ServiceProvider),Object(Klein\App),Object(Klein\Klein),Object(Klein\DataCollection\RouteCollection),Array)
#11 /var/www/html/sysPass/vendor/klein/klein/src/Klein/Klein.php(886): call_user_func(Object(Closure),Object(Klein\Request),Object(Klein\Response),Object(Klein\ServiceProvider),Object(Klein\App),Object(Klein\Klein),Object(Klein\DataCollection\RouteCollection),Array)
#12 /var/www/html/sysPass/vendor/klein/klein/src/Klein/Klein.php(588): Klein\Klein->handleRouteCallback(Object(Klein\Route),Object(Klein\DataCollection\RouteCollection),Array)
#13 /var/www/html/sysPass/lib/SP/Bootstrap.php(464): Klein\Klein->dispatch(Object(Klein\Request))
#14 /var/www/html/sysPass/lib/Base.php(75): SP\Bootstrap->run(Object(DI\Container))
#15 /var/www/html/sysPass/index.php(28): require(String)","caller":"N/A"}
```

**Comment by nuxsmin** [comment]
It should log some events in the event log (**WEB**).

**Comment by alexanderharm** [comment]
Sorry:

```
LDAP-Verbindungsfehler (BIND)
LDAP ERROR: Strong(er) authentication required (8)
LDAP DN: cn=syspass,cn=users,dc=ds,dc=example,dc=com
```

That brought me on the right track. So user-auth only works via 

- `ldaps://ldap.example.com:636` and `TLS` disabled
- `ldaps://ldap.example.com` and `TLS` disabled

All other combinations fail. `StartTLS` should work as well (does with OPNsense) so I guess there is a config issue somewhere.

In the WebGUI the following combinations work:

- `ldaps://ldap.example.com:636`, `TLS` disabled
- `ldaps://ldap.example.com`, `TLS` disabled
- `ldap://ldap.example.com:389`, `TLS` enabled
- `ldap://ldap.example.com`, `TLS` enabled
- `ldap.example.com:389`, `TLS` enabled
- `ldap.example.com`, `TLS` enabled

**Comment by nuxsmin** [comment]
Hello, StartTLS only works on unencrypted channels, since it adds an encrypted layer too, so somehow the other software is silently discarding this configuration, because TLS and ldaps can't work together...

Thanks for the feedback and testing!

**Comment by alexanderharm** [comment]
OK, understood. But why some configs work in the settings and not during login is a bit misleading...

**Comment by nuxsmin** [comment]
Sorry for the late reply.

When you check LDAP from settings it actually tries to bind using the connection user and not the user when loging in, but the connection parameters are the same.

Regards
